### PR TITLE
Added postgres user password option

### PR DESCRIPTION
**Pull request by MarcelCoding**
## Description of changes

Added option to enforce password of postrgres user.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/databases/postgresql.nix [hunk 1]
```diff
@@ -597,12 +603,26 @@ in
                   clauseSqlStatements = attrValues (mapAttrs (n: v: if v then n else "no${n}") filteredClauses);
 
                   userClauses = ''$PSQL -tAc 'ALTER ROLE "${user.name}" ${concatStringsSep " " clauseSqlStatements}' '';
+
+                  ensureUserPasswords = lib.optionalString
+                    (user.ensurePasswordFile != null)
+                    ''
+                      $PSQL -tA <<'EOF'
```
Comment (SuperSandro2000) [inline nixos/modules/services/databases/postgresql.nix:610]:
```suggestion
                  ensureUserPasswords = lib.optionalString (user.ensurePasswordFile != null) ''
                    $PSQL -tA <<'EOF'
```
we can keep the indentation a bit lower here

File: nixos/modules/services/databases/postgresql.nix [hunk 2]
```diff
@@ -597,12 +603,26 @@ in
...
+                        DO $$
+                        DECLARE password TEXT;
+                        BEGIN
+                          password := trim(both from replace(pg_read_file('${user.ensurePasswordFile}'), E'\n', '''));
+                          EXECUTE format('ALTER ROLE ${user.name} WITH PASSWORD '''%s''';', password);
```
Comment (SuperSandro2000) [inline nixos/modules/services/databases/postgresql.nix:615]: Do you know what happens here if the file is not readable or missing? Is that just ignored? Does the script abort?

**Review by SuperSandro2000 — COMMENTED** [review]
Please follow the [contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions) when naming your commits.

File: nixos/modules/services/databases/postgresql.nix (hunk 2 above)
Comment (MarcelCoding) [reply nixos/modules/services/databases/postgresql.nix:615]: The unit fails to start.

Comment (MarcelCoding) [reply nixos/modules/services/databases/postgresql.nix:615]: Example from `journalctl -efu postgresql`/`systemctl status postgresql`:

```
Dez 16 22:11:25 marcel-nix postgres[758943]: [758943] ERROR:  could not open file "/run/secrets/test" for reading: Permission denied
Dez 16 22:11:25 marcel-nix postgres[758943]: [758943] CONTEXT:  PL/pgSQL function inline_code_block line 4 at assignment
Dez 16 22:11:25 marcel-nix postgres[758943]: [758943] STATEMENT:  DO $$
Dez 16 22:11:25 marcel-nix postgres[758943]:           DECLARE password TEXT;
Dez 16 22:11:25 marcel-nix postgres[758943]:           BEGIN
Dez 16 22:11:25 marcel-nix postgres[758943]:             password := trim(both from replace(pg_read_file('/run/secrets/test'), E'\n', ''));
Dez 16 22:11:25 marcel-nix postgres[758943]:             EXECUTE format('ALTER ROLE test WITH PASSWORD ''%s'';', password);
Dez 16 22:11:25 marcel-nix postgres[758943]:           END $$;
Dez 16 22:11:25 marcel-nix postgresql-post-start[758942]: ERROR:  could not open file "/run/secrets/test" for reading: Permission denied
Dez 16 22:11:25 marcel-nix postgresql-post-start[758942]: CONTEXT:  PL/pgSQL function inline_code_block line 4 at assignment
```

File: nixos/modules/services/databases/postgresql.nix [hunk 3]
```diff
@@ -161,6 +161,12 @@ in
               '';
             };
 
+            ensurePasswordFile = mkOption {
+              type = lib.types.nullOr lib.types.str;
+              description = lib.mdDoc "Password of the user to ensure.";
```
Comment (h7x4) [inline nixos/modules/services/databases/postgresql.nix:166]:
```suggestion
              type = types.nullOr types.path;
              description = ''
                Path to a file containing the password of the user.
              '';
```

**Comment by MarcelCoding** [comment]
@h7x4 any further things?

**Comment by h7x4** [comment]
As far as I'm concerned, the diff looks good. A test would've been nice to ensure it's working properly, and so it doesn't suddenly break in the future.

That being said, I've seen there's been a lot of fuzz around the different postgres `ensure*` options lately, so I'd like some input from other people involved. Would you mind having a look at this? @RaitoBezarius @Ma27

**Review by Ma27 — COMMENTED** [review]
As stated back then, I think `ensure*` options are a bad idea and I'd like see less of it, not more of it, so I guess my opinion on that topic should be obvious.

**Comment by MarcelCoding** [comment]
Any alternative suggestions to provide passwords for postgres accounts?

**Comment by RaitoBezarius** [comment]
> Any alternative suggestions to provide passwords for postgres accounts?

You can use `preStart` of your service or spawn an ad-hoc systemd unit (oneshot, whatever you want) to provide passwords for Postgres accounts, there's _no_ need to put it _inside_ the PostgreSQL NixOS module.

In the future, when I have time, I will try to document and explain how people can build a set of ensure options for what they like _out of tree_.

**Comment by MarcelCoding** [comment]
That's my current workaround I use in production. I thought it would be a nice addition to the nixos module. But I did not know about the state of the ensure options.

Thanks for your Feedback!

**Comment by RaitoBezarius** [comment]
> That's my current workaround I use in production. I thought it would be a nice addition to the nixos module. But I did not know about the state of the ensure options.
> 
> Thanks for your Feedback!

It's not a workaround, it's a convergence option you are managing yourself and taking responsibility for.

Putting the convergence option inside of PostgreSQL and letting it unmaintained is the best way to have it shoot the foot of everyone on the long term, that's why we are pushing back.

Until we have better tools to tame the complexity of convergence, e.g. Terraform-style tooling, etc. We cannot reasonably offer those options beyond critical ones and a bunch of them which we can maintain.

**Comment by MarcelCoding** [comment]
All right.

**Pull request closed without merging**